onnxruntime/core/optimizer/qdq_transformer/qdq_final_cleanup.h

```diff
@@ -11,10 +11,15 @@ namespace onnxruntime {
 /**
     @Class QDQFinalCleanupTransformer
 
-    Remove any remaining back-to-back QuantizeLinear and DequantizeLinear pairs.
+    Remove any remaining back-to-back QuantizeLinear (Q) and DequantizeLinear (DQ) pairs.
+    This is the final Q/DQ cleanup. It should run AFTER other optimizers which rely on Q/DQ nodes, e.g., for fusion.
+    There are two cases:
 
-    This is the final cleanup where no quantized operator was available and we're going to run the operators
-    using fp32.
+    1. DQ -> Q
+    When the DQ and Q have the same scale and zero-point, they are always unnecessary and can be removed.
+
+    2. Q -> DQ
+    In this case, no quantized operator was available and we're going to run the operators using fp32.
```

Comment:
Naive question: I am guessing the model is invalid if we ever see a `Q->DQ` pair with different scale/zero point. If we ever saw this, shouldn't the final cleanup transformer report that as well and fail the model load process or are such sanity checks done elsewhere/ expected to be done somewhere else prior to trying to load the model (like the ONNX checker perhaps) already ?

Reply:
I'm not sure if this is checked elsewhere.
@yufenglee @skottmckay What do you think?

Reply:
I don't think the model makes sense if the zp and scale of adjacent DQ/Q nodes doesn't match. Technically the ONNX spec doesn't outlaw this, but wouldn't something be fundamentally broken elsewhere to create a model like that?